crypto/dilithium/p_dilithium3.c

```diff
@@ -9,117 +9,300 @@
 #include "../fipsmodule/evp/internal.h"
 #include "../evp_extra/internal.h"
 #include "sig_dilithium.h"
+#include "internal.h"
+#include "../fipsmodule/delocate.h"
 
-static int pkey_dilithium3_keygen(EVP_PKEY_CTX *ctx, EVP_PKEY *pkey) {
-  DILITHIUM3_KEY *key = OPENSSL_malloc(sizeof(DILITHIUM3_KEY));
-  if (key == NULL) {
-    goto err;
-  }
-  key->pub = OPENSSL_malloc(DILITHIUM3_PUBLIC_KEY_BYTES);
-  if (key->pub == NULL) {
-    goto err;
-  }
-  key->priv = OPENSSL_malloc(DILITHIUM3_PRIVATE_KEY_BYTES);
-  if (key->priv == NULL) {
-    goto err;
-  }
+// ML-DSA OIDs from as defined within:
+// https://csrc.nist.gov/projects/computer-security-objects-register/algorithm-registration
+//2.16.840.1.101.3.4.3.18
+static const uint8_t kOIDMLDSA65[]  = {0x60, 0x86, 0x48, 0x01, 0x65, 0x03, 0x04, 0x03, 0x12};
+
+// PQDSA functions: these are init/new/clear/free/get_sig functions for PQDSA_KEY
```

Comment:
I originally made the file move in this PR, but it made review difficult because all of the diffs no longer were comparing like for like changes. Andrew suggested making the file moves after the PR is merged to help review, but I can add back in the file changes if you'd prefer.

Essentially, those `PQDSA` functions will live in their own `p_pqdsa.c` file (within `fipsmodule/evp/p_pqdsa.c`) much like `p_kem.c` `p_ec.c` etc.

Reply:
I think if you move the file with `git mv` it should keep the diff as expected?

Reply:
Okay, now that we are deeper into review, I feel less apprehensive about file structure changes, as you are now more familiar with the review. I've made the file changes in [91056cb](https://github.com/aws/aws-lc/pull/1963/commits/91056cb9c81f7900894a59af2d2ac4a216e4da6a), note how github treats the re-names as new files, which is what I originally wanted to avoid. Nevertheless, the new file structure is as follows:


KEM | PQDSA | PQDSA NEW Location
-- | -- | --
crypto/fipsmodule/evp/p_kem.c | crypto/dilithium/p_dilithium3.c | crypto/dilithium/p_pqdsa.c
crypto/evp_extra/p_kem_asn1.c | crypto/dilithium/p_dilithium3_asn1.c | crypto/dilithium/p_pqdsa_asn1.c
crypto/fipsmodule/kem/kem.c | crypto/dilithium/p_dilithium3.c | crypto/dilithium/pqdsa.c
crypto/fipsmodule/ml_kem/ml_kem.c | crypto/dilithium/sig_dilithium.c | crypto/dilithium/mldsa.c
crypto/fipsmodule/ml_kem/ml_kem.h | crypto/dilithium/sig_dilithium.h | crypto/dilithium/mldsa.h

Then, when we eventually move to the fipsmodule, we'll have two directories:
- `crypto/fipsmodule/pqdsa/` which will contain generic `pqdsa` content, such as `pqdsa.c`
- `crypto/fipsmodule/ml_dsa/` which will contain `ml_dsa` specific content, such as `mldsa.c` `mldsa.h`
- `crypto/fipsmodule/evp` can contain `p_pqdsa.c`
- `crypto/evp_extra` can contain `p_pqdsa_asn1.c`